### PR TITLE
Use mut slices

### PR DESCRIPTION
**Pull request by axic**
Part of #24.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -472,29 +472,30 @@ pub fn create(value: &EtherValue, data: &[u8]) -> CreateResult {
...
-
-    ret
 }
 
 /// Returns a vector containing all data passed with the currently executing call.
 pub fn calldata_acquire() -> Vec<u8> {
-    unsafe_calldata_copy(0, calldata_size())
+    let length = calldata_size();
+
+    let mut ret: Vec<u8> = unsafe_alloc_buffer(length);
```
Comment (jakelang) [inline src/lib.rs]: What is the purpose of using `unsafe_alloc_buffer`? It seems like you can do this safely using `From` or `slice.to_vec()`

**Comment by jakelang** [comment]
Needs to be rebased.

File: src/lib.rs (hunk 1 above)
Comment (axic) [reply src/lib.rs]: Uhm, have you seen the code? :)

Comment (jakelang) [reply src/lib.rs]: that's not exactly a very effective way of pointing out the problem.

Comment (axic) [reply src/lib.rs]: After the change it is only used by the `_acquire` functions. By requiring them to use a slice defeats their purpose, one could just use the underlying functions.

I'm wondering if we keep the `_acquire` ones in with `std::Vec`, but the caller doesn't use it, will Rust properly optimise `std::Vec` out?

Comment (axic) [reply src/lib.rs]: Then again we could make it depend on `#feature(!no_std)`

**Comment by axic** [comment]
Please don't merge, not all of them are translated yet.

File: src/lib.rs [hunk 2]
```diff
@@ -472,29 +472,30 @@ pub fn create(value: &EtherValue, data: &[u8]) -> CreateResult {
...
+    ret
 }
 
 /// Returns the segment of call data beginning at `from`, and continuing for `length` bytes.
-pub fn calldata_copy(from: usize, length: usize) -> Result<Vec<u8>, Error> {
+pub fn calldata_copy(from: usize, length: usize, ret: &mut [u8]) -> Result<(), Error> {
     let size = calldata_size();
 
-    if (size < from) || ((size - from) < length) {
+    if (size < from) || ((size - from) < length || ret.len() < length) {
```
Comment (axic) [inline src/lib.rs]: Should this require an exact slice?

File: src/lib.rs [hunk 3]
```diff
@@ -472,29 +472,30 @@ pub fn create(value: &EtherValue, data: &[u8]) -> CreateResult {
...
+    let length = calldata_size();
+
+    let mut ret: Vec<u8> = unsafe_alloc_buffer(length);
+    unsafe_calldata_copy(0, length, &mut ret);
+    ret
 }
 
 /// Returns the segment of call data beginning at `from`, and continuing for `length` bytes.
-pub fn calldata_copy(from: usize, length: usize) -> Result<Vec<u8>, Error> {
+pub fn calldata_copy(from: usize, length: usize, ret: &mut [u8]) -> Result<(), Error> {
```
Comment (axic) [inline src/lib.rs]: Should the parameter order be `from, ret, length` or `from, length, ret`?

File: src/lib.rs (hunk 2 above)
Comment (jakelang) [reply src/lib.rs]: not sure what you mean. all slices have an associated fixed length

File: src/lib.rs (hunk 3 above)
Comment (jakelang) [reply src/lib.rs]: `ret, from, len` would be preferred for me, but most people would probably like your first option. Mine is just more in line with Intel ASM syntax and git push, which is why I personally think it is ergonomic.

**Comment by jakelang** [comment]
Need to run fmt on this.

File: src/lib.rs (hunk 2 above)
Comment (axic) [reply src/lib.rs]: Options
a) `ret.len() == length` or
b) good as long as the slice is bigger than what we want to copy (current implementation is this)

Comment (jakelang) [reply src/lib.rs]: If we have separate `calldata_acquire` method then a) is the obvious option. the question is what use cases a) would have, and whether they would be very ergonomic at this layer of abstraction.

**Comment by jakelang** [comment]
@axic is this ready for merge?

**Comment by axic** [comment]
I think we never made a decision on https://github.com/ewasm/ewasm-rust-api/pull/28#discussion_r239648057, but can merge it without that.

I think it should be ready.

File: src/lib.rs (hunk 2 above)
Comment (axic) [reply src/lib.rs]: Ah right this question does needs an answer. `ret.len() == length` or `ret.len() >= length` should be the condition.

**Comment by jakelang** [comment]
@axic I gave my opinion in the comments, you can make the decision

File: src/lib.rs (hunk 2 above)
Comment (jakelang) [reply src/lib.rs]: second one would be unsafe if the same buffer is used to write multiple times, because underflow is possible.

Comment (axic) [reply src/lib.rs]: Hm? These are slices, there's no overflow. The assertion is `ret.len() <= length` or `ret.len() == length`.

**Pull request merged**